### PR TITLE
[Import] Handle removed specs

### PR DESCRIPTION
**Pull request by segiddins**
I don't know how to create new fixtures, hence the lack of tests

### CODE REVIEW

**Comment by segiddins** [comment]
\c @floere

**Comment by floere** [comment]
We also need to

* [ ] Add a spec.
* [ ] Trigger various remove webhooks.

**Comment by floere** [comment]
@segiddins Are you insisting that the pod version be removed? Alternative suggestion: Mark version as deleted, add commit to deleted version. How's that sound?

**Comment by segiddins** [comment]
@floere that would prevent re-adding that version again in the future though.

**Comment by floere** [comment]
@segiddins How so?

**Comment by segiddins** [comment]
Because the version would already exist when you go to create it?

**Comment by floere** [comment]
For an external viewer, the version does not exist. Internally, it is just marked as deleted.




So re-adding a version, for an external viewer, is creating the version. Internally, we un-delete the version.




This internal/external distinction is necessary if we'd like to use a deleted flag. It adds a bit of complexity when thinking about it, but is worth it as we can track all commits.




An example of an external service is search which ignores deleted versions (and pods).

**Comment by floere** [comment]
@segiddins Does this make sense to you?

**Comment by segiddins** [comment]
Yup, makes sense!

**Comment by segiddins** [comment]
So Ion this more, and it just doesn't work, because when we go to fetch the spec, it will be nil.

**Comment by segiddins** [comment]
@floere I don't love this change, but I think it makes things work

**Comment by orta** [comment]
commenting to remind myself to start adding checks for deletion in cocoapods.org

File: app/models/commit/import.rb [hunk 1]
```diff
@@ -42,7 +42,11 @@ def self.import(commit_sha, type, files, committer_email, committer_name)
           files.each do |file|
             next unless file =~ /\.podspec(.json)?\z/
 
+            _, name, version, = file.split(File::SEPARATOR)
+            stub_spec = ::Pod::Specification.new(nil, name) { |s| s.version = version }
+
             spec = fetch_spec(commit_sha, file)
+            spec = stub_spec if type == :removed
```
Comment (floere) [inline app/models/commit/import.rb]: We first fetch it, then we stub it? Should they not exclude each other?

File: app/models/commit/import.rb [hunk 2]
```diff
@@ -42,7 +42,11 @@ def self.import(commit_sha, type, files, committer_email, committer_name)
           files.each do |file|
             next unless file =~ /\.podspec(.json)?\z/
 
+            _, name, version, = file.split(File::SEPARATOR)
+            stub_spec = ::Pod::Specification.new(nil, name) { |s| s.version = version }
```
Comment (floere) [inline app/models/commit/import.rb]: The stub is created in any case? Is this necessary?

**Comment by segiddins** [comment]
@orta it's the same published? Check that's used elsewhere in trunk

**Comment by floere** [comment]
I've rewritten it a bit. Two specs still fail, but it's a sign they were iffy in the first place and need improvement.

**Comment by floere** [comment]
@segiddins Are you happy with my changes?

The things that remain:

* [ ] Trigger the deleted webhook.
* [ ] Decide what happens when all versions of a pod are deleted. Is the pod deleted?

I'd say yes to my latter question. Agreed?

**Comment by floere** [comment]
I've added code/a test which adds functionality for "deleting" a pod if all its versions are "deleted". I think this is the way to go. A versionless pod cannot exist in the CP world.

**Comment by floere** [comment]
I'm quite happy with this PR now – I hope I have addressed your concerns, @segiddins?

File: app/models/commit/import.rb [hunk 3]
```diff
@@ -8,57 +8,69 @@
...
+              handle_with_existing_spec(type, spec, pod, committer, commit_sha)
+            end
           end
         end
 
+        # Extracts the pod name and the version name from the file name.
+        #
+        def extract_name_and_version(file_name)
+          _, name, version_name, _ = *file_name
+            .match(%r{([^\/]+)\/([^\/]+)\/[^\.]+\.podspec(.json)?\z})
```
Comment (segiddins) [inline app/models/commit/import.rb]: why is this better than just using split and taking results 2 and 3?

File: app/models/commit/import.rb [hunk 4]
```diff
@@ -66,32 +78,128 @@ def self.handle_modified(spec, pod, committer, commit_sha)
...
         end
+
+        # If we have a version for the given pod and spec, we remove it.
+        #
+        # @param spec [Pod::Specification] The removed podspec.
+        # @param pod [Pod] The removed version's pod.
+        # @param committer [Owner] The committer.
+        # @param commit_sha [String] The git commit SHA-1.
+        #
+        # TODO: Needs big fat logging (an informative log message on version).
```
Comment (segiddins) [inline app/models/commit/import.rb]: let's not use the phrase `big fat`, please?

File: app/models/commit/import.rb (hunk 3 above)
Comment (floere) [inline app/models/commit/import.rb]: Because it attaches to the end.

File: app/models/commit/import.rb (hunk 4 above)
Comment (floere) [inline app/models/commit/import.rb]: `very large` is of course fine too.

Comment (floere) [inline app/models/commit/import.rb]: I've addressed this. Thanks!

File: app/models/commit/import.rb (hunk 3 above)
Comment (floere) [inline app/models/commit/import.rb]: What I meant is (it was late): we check for the podfile by asserting against the extensions. We can use the results for this assertion further down. I also ran into trouble with the tests.

I'd say feel free to change this back! :)

**Comment by floere** [comment]
After hearing feedback from Olivier Halligon…

This PR can only be pulled if these additional features are added:

* [ ] Undelete a pod if it is pushed.
* [ ] Undelete a pod version if it is pushed.
* [ ] Undelete a pod if a pod version is undeleted where before there were only deleted versions.

**Comment by segiddins** [comment]
Versions are already marked undeleted when a new one is pushed

**Comment by floere** [comment]
@segiddins The new one? Yes. What about a version that is re-pushed? Currently, Trunk will do nothing:

```
          if version = pod.versions_dataset.where(:name => specification.version).first
            if version.published?
              headers 'Location' => url(version.resource_path)
              message = "Unable to accept duplicate entry for: #{specification}"
              json_error(409, message)
            end
          else
            version = pod.add_version(:name => specification.version)
          end
```

**Comment by floere** [comment]
Also:

* [ ] Only push to GitHub if the version/pod did not exist yet (or e.g. was not deleted).

Actually keeping everything in sync is a bit of work.

**Comment by segiddins** [comment]
```ruby
def push!(committer, specification_data)
  update(:deleted => false)
```

**Comment by floere** [comment]
@segiddins Oh, thanks!

* [ ] What happens if a commit comes in via HooksController (re-adding an existing version) and then goes back out to GitHub?

**Comment by segiddins** [comment]
```
+          commit = first_or_add_commit(version, commit_sha, spec, committer)
+          version.update(:deleted => false)
```

**Comment by floere** [comment]
Actually, in this case:

```
def push!(committer, specification_data)
  update(:deleted => false)
  response = PushJob.new(self, committer, specification_data).push!
  if response.success?
    add_commit(:committer => committer, :sha => response.commit_sha, :specification_data => specification_data)
    pod.add_owner(committer) if pod.owners.empty?
  end
  response
end
```

Seems to work fine if a version is re-added. It simply is undeleted and gets a new commit.

**Comment by floere** [comment]
So my task list is reduced to:

* [ ] Undelete a pod if it is pushed (in `pods_controller` - rewrite `Pod.create`, use a new method that handles the `delete` case or creates a new pod).
* [ ] Undelete a pod if a pod version is undeleted where before there were only deleted versions (count undeleted versions after updating a version `deleted` attribute in a central spot, and undelete a pod if necessary).

**Comment by floere** [comment]
Writing a mini-spec for the deleted attribute:

* The deleted attribute is CP internal. External there is only the existence/non-existence of a pod, a version.
* If a version is created, we create or undelete it.
* If the last undeleted version of a pod is deleted, the pod will be deleted.
* As soon as an undeleted version of a deleted pod exists, the pod is undeleted.

@segiddins Is that enough to define the lifecycle?

**Comment by segiddins** [comment]
@floere correct

**Comment by floere** [comment]
@segiddins Thank you! :rocket:

**Comment by segiddins** [comment]
What's left to do on this one?

**Comment by floere** [comment]
@segiddins Get the specs to run, perhaps? Otherwise I think it's fine. I've started work on the lifecycler mentioned in https://github.com/CocoaPods/trunk.cocoapods.org/commit/6533e64bf374b095a7acc54620769314fc6211ef, but we could continue work on that after this issue and pull this one in.

**Comment by segiddins** [comment]
@floere I think it will pass if the pod_versions.deleted column is added?

**Comment by floere** [comment]
@segiddins True – that means we should add it in Humus before this can get pulled.

**Comment by orta** [comment]
Looks like the local db:migrations are now pretty out of sync with humus migrations, do we think it's worth adding humus as a submodule to run rake tasks in?

**Comment by floere** [comment]
@orta What do you mean by `local db:migrations` – trunk local?

**Comment by orta** [comment]
Trunk's migrations: https://github.com/CocoaPods/trunk.cocoapods.org/tree/master/db/migrations
Humus's Trunk Migrations: https://github.com/CocoaPods/Humus/tree/master/migrations/trunk

Humus' is the"one true state" here.

**Comment by floere** [comment]
We need to remove the trunk local migrations. I've incorporated them in Humus. Shall I remove them or will you?

**Comment by orta** [comment]
You please, I'm about to have 2 hours of meetings, I have just enough time to ship something else

**Comment by floere** [comment]
Ok, and thanks for checking! :)

**Comment by orta** [comment]
This is now green!

**Comment by floere** [comment]
@orta Thanks!
@segiddins Let's give it one more look before merging…

File: .travis.yml [hunk 5]
```diff
@@ -1,4 +1,11 @@
...
-- env RACK_ENV=test bundle exec rake db:bootstrap
+before_script: .travis/before.sh
 install: bundle install --deployment --without development production
+rvm: 2.1.3
+cache: bundle
+addons:
+  postgresql: "9.4"
+  apt:
+    packages:
+    - postgresql-server-dev-9.4
```
Comment (segiddins) [inline .travis.yml]: missing trailing newline

**Comment by segiddins** [comment]
Let's :ship: ? I'll be by my computer for a little while to moniter

**Comment by orta** [comment]
Works for me. Will deploy.

**Pull request merged**